Support for default request headers (#461)
reactphp/http: README.md

method actually returns a *new* [`Browser`](#browser) instance with the
given setting applied.

#### withHeader()

The `withHeader(string $header, string $value): Browser` method can be used to
add a request header for all following requests.

> Note that the new header will overwrite any headers previously set with the same name (case-insensitive). Following requests will use these headers by default unless they are explicitly set for any requests.

Reviewer: Can you add a example in here, could be something like this:
```php
$browser = $browser->withHeader('User-Agent', 'ACME');

$browser->get($url)->then(function (Psr\Http\Message\ResponseInterface $response) {
    var_dump($response);
}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;
});
```
Same goes for the `withoutHeader()` method and the two docblocks in `Browser.php`.

Reviewer 2: added example code to the `withHeader()` and `withoutHeader()` in the README

Reviewer: @51imyy Nice, can you also add them to the docblocks.

Reviewer 2: Added the examples to the docblocks.

#### withoutHeader()

The `withoutHeader(string $header): Browser` method can be used to
remove any default request headers previously set via the [`withHeader()` method](#withheader).

> Note that this method only affects the headers which where set with the method `withHeader(string $header, string $value): Browser`

Reviewer: Little typo
```suggestion
Note that this method only affects the headers which were set with the method `withHeader(string $header, string $value): Browser`
```

method actually returns a *new* [`Browser`](#browser) instance with the
given setting applied.

#### withHeader()

The `withHeader(string $header, string $value): Browser` method can be used to
add a request header for all following requests.

```php
$browser = $browser->withHeader('User-Agent', 'ACME');

$browser->get($url)->then(function (Psr\Http\Message\ResponseInterface $response) {
    //will contain the header User-Agent with the value ACME
    var_dump($response->getHeaders());
}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;
});
```

> Note that the new header will overwrite any headers previously set with the same name (case-insensitive). Following requests will use these headers by default unless they are explicitly set for any requests.

Reviewer: @51imyy Can you break this line down ^^

Reviewer 2: broke the line down.

#### withoutHeader()

The `withoutHeader(string $header): Browser` method can be used to
remove any default request headers previously set via the [`withHeader()` method](#withheader).

```php
$browser = $browser->withoutHeader('User-Agent');

$browser->get($url)->then(function (Psr\Http\Message\ResponseInterface $response) {
    //won't contain the User-Agent header
    var_dump($response->getHeaders());
}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;
});
```

> Note that this method only affects the headers which were set with the method `withHeader(string $header, string $value): Browser`

Reviewer: @51imyy Can you break this line down ^^

Reviewer 2: brkoke the line down.

method actually returns a *new* [`Browser`](#browser) instance with the
given setting applied.

#### withHeader()

The `withHeader(string $header, string $value): Browser` method can be used to
add a request header for all following requests.

> ```php

Reviewer: Can you add an empty line between these two (you already did this in the `withoutHeader()` description)

Reviewer 2: added a new line. :D

method actually returns a *new* [`Browser`](#browser) instance with the
given setting applied.

#### withHeader()

The `withHeader(string $header, string $value): Browser` method can be used to
add a request header for all following requests.

> ```php
> $browser = $browser->withHeader('User-Agent', 'ACME');
> 
> $browser->get($url)->then(function (Psr\Http\Message\ResponseInterface $response) {
>     //will contain the header User-Agent with the value ACME
>     var_dump($response->getHeaders());

Reviewer: I think it can be confusing that we show two types of headers in this example (request and response).
```suggestion
    var_dump($response);
```

}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;

> });

Reviewer 3: ```suggestion
$browser->get($url)->then(…);
```

```

Note that the new header will overwrite any headers previously set with
the same name (case-insensitive). Following requests will use these headers
by default unless they are explicitly set for any requests.

#### withoutHeader()

The `withoutHeader(string $header): Browser` method can be used to
remove any default request headers previously set via
the [`withHeader()` method](#withheader).

```php

> $browser = $browser->withoutHeader('User-Agent');
> 
> $browser->get($url)->then(function (Psr\Http\Message\ResponseInterface $response) {
>     //won't contain the User-Agent header
>     var_dump($response->getHeaders());

Reviewer: I think it can be confusing that we show two types of headers in this example (request and response).
```suggestion
    var_dump($response);
```

}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;

> });

Reviewer 3: ```suggestion
$browser->get($url)->then(…);
```

method actually returns a *new* [`Browser`](#browser) instance with the
given setting applied.

> 

Reviewer 3: nit: whitespace

Reviewer 2: i removed the whitespace

#### withHeader()

The `withHeader(string $header, string $value): Browser` method can be used to
add another header.

> For example:

Reviewer 3: ```suggestion
The `withHeader(string $header, string $value): Browser` method can be used to
add a request header for all following requests.
```

```php
$browser = $browser->withHeader('User-Agent', 'ACMC');

$browser->get($url)->then(function (Psr\Http\Message\ResponseInterface $response) {

>     // case insensitiv

Reviewer 4: ```suggestion
    // case-insensitive
```

>     var_dump($response->getHeader('user-agent'));

Reviewer 3: Note sure how useful this example is as the `withHeader()` method would set an outgoing request header while this dumps the incoming response headers here?

Reviewer 2: I removed all the code examples for the methods "withHeader()" and "withoutHeader()"

}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;
});
```

> Note that the new header will (case insensitiv) overwrite an existing header with the same name.

Reviewer 4: ```suggestion
Note that the new header will (case-insensitive) overwrite an existing header with the same name.
```

> Headers passed to the HTTP methods will be preferred.

Reviewer 3: ```suggestion
Note that the new header will overwrite any headers previously set with the same name (case-insensitive). Following requests will use these headers by default unless they are explicitly set for any requests.
```

```php
$browser = $browser->withHeader('User-Agent', 'ACMC');

$browser->get($url, array('user-agent' => 'value'))->then(function (Psr\Http\Message\ResponseInterface $response) {

>     // case insensitiv

Reviewer 4: ```suggestion
    // case-insensitive
```

    var_dump($response->getHeaders()); 
    //will return array(
    //                    ...
    //                    'user-agent' => 'value'
    //                    ...                                      
    //                 )
}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;
});
```

> 

Reviewer 3: I see why this is added, but I think this example might be a bit confusing and less useful for the average consumer. My suggestion would be to drop this (KISS).

Reviewer 2: dropped it

#### withoutHeader()

The `withoutHeader(string $header): Browser` method can be used to remove an added header
(which got added via the `withHeader(string $header, string $value): Browser`).

> This works case insensitiv.

Reviewer 4: ```suggestion
This works case-insensitive.
```

> The `withoutHeader(string $header): Browser` method can be used to remove an added header
> (which got added via the `withHeader(string $header, string $value): Browser`).
> This works case-insensitive.
> It won't remove headers which will be passed directly into the HTTP method
> For example:

Reviewer 3: ```suggestion
The `withoutHeader(string $header): Browser` method can be used to
remove any default request headers previously set via the [`withHeader()` method](#withheader).
```

```php
$browser = $browser->withHeader('User-Agent', 'ACMC');
$browser = $browser->withoutHeader('USER-AGENT');

$browser->get($url)->then(function (Psr\Http\Message\ResponseInterface $response) {

>     // case insensitiv

Reviewer 4: ```suggestion
    // case-insensitive
```

>     var_dump($response->getHeader('User-Agent')); // will return empty array

Reviewer 3: Same problem as above, this is dumping a response header which is unlikely to exist.

}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;
});

``````php
$browser = $browser->withoutHeader('USER-AGENT');

$header = array('User-Agent' => 'ACMC'); // won't be affected by the method `withoutHeader(string $header): Browser`
$browser->get($url, $header)->then(function (Psr\Http\Message\ResponseInterface $response) {

>     // case insensitiv

Reviewer 4: ```suggestion
    // case-insensitive
```

    var_dump($response->getHeader('User-Agent'));
}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;
});

> ```

Reviewer 3: Should we remove this like the one above?

> Note that this method only affects the headers which where set with the method `withHeader(string $header, string $value): Browser`

Reviewer 3: Note that this will remove any headers previously set with the same name (case-insensitive). Following requests will no longer use these headers by default unless they are explicitly set for any requests.
